client/my-sites/earn/components/add-edit-coupon-modal/test/index.jsx

```diff
@@ -0,0 +1,353 @@
+		);
+		getconnectedAccountMinimumCurrencyForSiteId.mockImplementation(
+			mockGetConnectedAccountMinimumCurrencyForSiteId
+		);
+
+		modalRoot = document.createElement( 'div' );
+		modalRoot.setAttribute( 'id', 'wpcom' );
+		document.body.appendChild( modalRoot );
+		Modal.setAppElement( modalRoot );
+	} );
+
+	afterEach( () => {
+		unmountComponentAtNode( modalRoot );
+		document.body.removeChild( modalRoot );
+		[ ...document.getElementsByClassName( 'ReactModalPortal' ) ].map( ( el ) => {
+			document.body.removeChild( el );
+		} );
+		modalRoot = null;
+		global.Date = RealDate;
+	} );
+
+	test( 'should render empty form when no coupon is provided', () => {
+		mockDate( '2023-06-23T14:23:44z' );
+		render(
+			<WrappedRecurringPaymentsCouponAddEditModal closeDialog={ closeDialog } coupon={ {} } />
+		);
+		const codeTextbox = getCodeTextbox();
+		const amountOption = getAmountOption();
+		const percentageOption = getPercentageOption();
+		const amountInput = getAmountInput();
+		const limitDurationToggle = getLimitDurationToggle();
+		const limitDurationSelection = getLimitDurationSelection();
+		const limitCouponsToggle = getLimitCouponsToggle();
+		const limitCouponsTextInput = getLimitCouponsTextInput();
+
+		expect( codeTextbox.value ).toEqual( '' );
+		expect( amountOption.selected ).toBe( false );
+		expect( percentageOption.selected ).toBe( true );
+		expect( amountInput ).name = 'discount_percentage';
+		expect( amountInput ).type = 'text';
+		expect( amountInput ).value = '';
+		expect( limitDurationToggle ).not.toBeChecked();
+		expect( limitDurationSelection ).toBeDisabled();
+		expect( limitCouponsToggle ).not.toBeChecked();
+		expect( limitCouponsTextInput ).toBeDisabled();
+	} );
+
+	test( 'should generate a new coupon code when the random button is clicked', () => {
+		mockDate( '2023-06-23T14:23:44z' );
+		render(
+			<WrappedRecurringPaymentsCouponAddEditModal closeDialog={ closeDialog } coupon={ {} } />
+		);
+		const randomButton = getRandomButton();
+		const codeTextbox = getCodeTextbox();
+
+		expect( codeTextbox.value ).toEqual( '' );
+		waitFor( () => randomButton.click() );
+		waitFor( () => {
+			const generatedCode = codeTextbox.value;
+			expect( generatedCode ).not.toEqual( '' );
+			expect( generatedCode.length ).toBeGreaterThan( 3 );
+		} );
+	} );
+
+	test( 'should show discount amount input when discount type is amount', async () => {
+		mockDate( '2023-06-23T14:23:44z' );
+		render(
+			<WrappedRecurringPaymentsCouponAddEditModal closeDialog={ closeDialog } coupon={ {} } />
+		);
+		const discountTypeSelect = getDiscountTypeSelect();
+		const amountOption = getAmountOption();
+		const percentageOption = getPercentageOption();
+		const amountInput = getAmountInput();
+
+		expect( amountOption.selected ).toBe( false );
+		expect( percentageOption.selected ).toBe( true );
+		expect( amountInput ).name = 'discount_percentage';
+		expect( amountInput ).type = 'text';
+		expect( amountInput ).value = '';
```

Comment:
Nit, but in case `@testing-library/jest-dom` was available for the tests here, we could use some of those assertions to have the tests read better (eg. [`toHaveValue `](https://github.com/testing-library/jest-dom#tohavevalue), [`toHaveAttribute`](https://github.com/testing-library/jest-dom#tohaveattribute)).

More in general, though, it's always better to write tests from the point of view of the user. A user wouldn't go and check the `name`, `type`, and `value` attributes of the amount input. Instead, for the user it's important that such an input is displayed. Therefore, an alternative way to to write the last part of the test could be to

`expect( screen.getByRole('input`, {name: 'Accessible name to the user'})).toHaveValue('')`


The same would apply for other assertions later in the file.

Related read: https://kentcdodds.com/blog/common-mistakes-with-react-testing-library

Reply:
Thanks for this. I've refactored these tests accordingly. I appreciate the link as well, that's a handy reference and it cleared up a few mistakes I was making!